fonz/connection.py

```diff
@@ -146,100 +255,121 @@ def get_dimensions(self, model: str, explore_name: str) -> List[str]:
+        async with session.get(url=url) as response:
+            if response.status == 204:
+                logger.debug(f"Query task {query_task_id} not finished yet.")
+                raise QueryNotFinished
+            logger.debug(f"Received results from query {query_task_id}")
+            result = await response.json()
+            return result
+
+    async def query_explore(self, model: Model, explore: Explore):
+        async with aiohttp.ClientSession(
+            headers=self.session.headers, raise_for_status=True
+        ) as async_session:
+            dimensions = [dimension.name for dimension in explore.dimensions]
+            query_id = await self.create_query(
+                async_session, model.name, explore.name, dimensions
+            )
+            explore.query_id = query_id
+            query_task_id = await self.run_query(async_session, query_id)
+            result = await self.get_query_results(
+                async_session, query_task_id, explore.name
             )
-        query_result = response.json()
 
-        return query_result
+            if not result:
+                return
+            elif isinstance(result, dict) and result.get("errors"):
+                first_error = result["errors"][0]
+                error_message = first_error["message_details"]
+                line_number = first_error["sql_error_loc"]["line"]
+                sql = result["sql"]
 
-    def get_query_sql(self, query_id: int) -> str:
-        """Collect the SQL string for a Looker query."""
+                explore.errored = True
+                model.errored = True
+                explore.error = SqlError(error_message, sql, line_number)
 
-        logger.debug(f"Getting SQL for query {query_id}")
-        url = utils.compose_url(self.base_url, path=["queries", query_id, "run", "sql"])
-        response = self.session.get(url=url)
-        try:
-            response.raise_for_status()
-        except requests.exceptions.HTTPError as error:
-            raise FonzException(
-                f'Failed to obtain SQL for query "{query_id}".\nError raised: "{error}"'
+    async def query_dimension(
+        self, model: Model, explore: Explore, dimension: Dimension
+    ):
+        async with aiohttp.ClientSession(
+            headers=self.session.headers, raise_for_status=True
+        ) as async_session:
+            query_id = await self.create_query(
+                async_session, model.name, explore.name, [dimension.name]
+            )
+            dimension.query_id = query_id
+            query_task_id = await self.run_query(async_session, query_id)
+            result = await self.get_query_results(
+                async_session, query_task_id, explore.name
             )
-        sql = response.text
 
-        return sql
+            if not result:
+                return
+            elif isinstance(result, dict) and result.get("errors"):
+                first_error = result["errors"][0]
```

Comment:
Is this `first_error` in that there are multiple errors or `first_error` in that the error in in the first 'row'?

Reply:
It's the first error in the array of errors returned by the API. The API response for the Async Query Result endpoint seems a little half-baked, but it seems you can get multiple errors back from the endpoint.